[#155] Overhaul of InitializerMessageSource.

mseaton opened the pull request:
This PR undertakes a reasonably sizeable overhaul of the InitializerMessageSource.  The initial impetus for this was to investigate and fix issues related to appropriate "fallback" messages not being rendered when expected.  For example, we had situations where message codes were being rendered even when messages exist for the "default locale" (i.e. English).  Upon starting to investigate, we encountered a number of other examples of unexpected behavior, which we attributed to limitations in the existing OpenMRS and Spring message source implementations around precedence of message loading in both Spring and Openmrs message sources, and our delegation of message loading between custom source and parent source.

Ultimately, we concluded that the most effective means of ensuring consistent, expected behavior was to no longer split message resolution responsibility between custom source and parent source, but to fully implement a new message source to cover all known use cases.  Thus, this new message source is a replacement for the OpenMRS message source, not a child to it.  It takes reponsibility for loading messages first from the OpenMRS core classpath, then from each started module classpath, and finally from Initializer domains.

This new message source also adds in improved support for fallback locales, using the OpenMRS LocaleUtility.getLocalesInOrder() function to fallback to configured default and supported locales, before finally falling back to the system locale.  The new message source also allows for other language-specific fallbacks to be configured as needed - i.e. one can now configure that the "ht" Locale should use "fr" as a fallback language before falling back as described above.

@mks-d, @Ruhanga, @mogoodrich, @rbuisson - please let me know your thoughts and how we can move this forward.

Ruhanga reviewed (commented):
@mseaton, thanks a bunch for raising this PR.  Would you also add a `readme` note highlighting this enhancement.

mseaton commented on api/src/main/java/org/openmrs/module/initializer/InitializerMessageSource.java:
```diff
@@ -50,104 +32,227 @@
+	/**
+	 * @return an array of message property resource file names found on the classpath
+	 */
+	protected Map<String, Properties> getMessagePropertyResourcesFromClasspath() {
```
@mks-d no, the details of how this is done in core are buried in private methods within the core message source (org.openmrs.messagesource.impl.MutableResourceBundleMessageSource).  I tried very hard to build on this core message source or, failing that, to build on the ReloadableResourceBundleMessageSource that the core message source extends and which is provided by Spring.  In the end, I was unable to get all of my unit tests to pass using these techniques and came to the conclusion that it was not possible.

In the existing code, the InitializerMessageSource adds the core message source as a parent, which ends up delegating to that message source if the InitializerMessageSource does not resolve a message.  The way I understand this to work is that:

a) The InitializerMessageSource will look for matches for a code in a Locale.  If this returns null, then
b) The parent will be delegated to, to look for matches for a code in a Locale

Now, the way the parent works (from my understanding) is that it is given an ordered array of resource bundle locations to check (i.e. a directory that has message property files it it for various locales).  It will check each in turn and if it finds a matching message, it will return it, without checking later resources in the array.  What this means is that if you are looking for a message for "fr_FR", and it finds a message for "fr" in the resource bundle at array index 1, it will return it, even if a better matching message is defined in "fr_FR" in the resource bundle at array index 2.  It's not really possible to ask the parent for the message at a specific locale and work with this.  You can only ask the parent to resolve a message a given locale and have it return a match at this (and less specific) locales.

Ultimately, I think there are issues like this that I would consider bugs in both the Initializer Message Source and _also_ in the OpenMRS core message source, and I wanted to address all of these.  So, in this new Message Source, I am basically implementing what I think a new OpenMRS message source should look like, that would handle all of the core needs and allow extension by Initializer.  Ultimately if this is successful and after an iteration or two of additional features, etc. I would encourage us to harvest this back into core as a replacement.

mseaton replied:
Beyond this, there are things that I am doing in this message source that core is not, and likely should be, around which specific message properties files are loaded from the classpath.  I suspect that core is loading all of the message properties that are in module dependencies as well (including multiple versions of core).  I think if we were to write a collection of unit tests for just the core message source, with expected behaviors around how module messages should override core messages, or each other, and how these should handle messages with variant, country, and language options, or with fallback locales, then we will find that there are a number of these tests that will fail with core.

mseaton commented:
> @mseaton, thanks a bunch for raising this PR. Would you also add a `readme` note highlighting this enhancement.

@Ruhanga - yes, if helpful.  It wasn't 100% clear to me what this should look like.  I'd consider this more of a bug fix than an enhancement at this stage.  @mks-d ?

mogoodrich commented:
@mseaton haven't had a chance to review detail yet, will look at this tomorrow, but re: "namely that one should be able to configure the "ht" Locale to use "fr" as a fallback language before finally falling back to the system locale"...  I think that was originally working with the main OpenMRS Message Source by setting the GP "default_locale" to "fr".  It sounds like what you have adds is more flexible so worth keeping, but we might want to include the "default_locale" GP in the fallback chain (if you haven't already), ie after the language-specific fallback but before the system locale fallback.

mseaton commented:
> I think that was originally working with the main OpenMRS Message Source

As I wrote above, I do think this is a nice feature to add in @mogoodrich , but I haven't yet found any evidence that OpenMRS does this in the core message source.  I'd be very interested if you could reproduce this or point me to code that is in place that does this.

Note, if we were to implement this here, the static OpenMRS methods: LocaleUtility.getDefaultLocale() and LocaleUtility.getLocalesInOrder() are likely useful.

mogoodrich commented:
Interesting @mseaton ... I haven't looked at the code, but I thought this was working at some point, and was how we had fallback from Kreyol to French working (French is our default language in Haiti).  However, as this fallback no longer seems to be working, it's entirely possible that this was a feature in an earlier version of OpenMRS that was broken or removed at some point.  If your exploratio shows that the current message source doesn't do this, I'm fine with skipping.

mogoodrich commented on api/src/test/java/org/openmrs/module/initializer/InitializerMessageSourceIntegrationTest.java:
```diff
@@ -9,79 +9,128 @@
+	public void shouldLoadTranslationsWithAppropriateFallbacks() {
+		Context.setLocale(HAITIAN_KREYOL);
+		testMessage("Only In ht", "messageOnlyInHt"); // Direct match
+		testMessage("messageOnlyInFrFr", "messageOnlyInFrFr"); // Fallback to message code
```
Super nit, but comment should probably say something like "no match, fallback to message code"... I was slightly confused at first

mseaton replied:
Done

mogoodrich commented on api/src/main/java/org/openmrs/module/initializer/InitializerMessageSource.java:
```diff
@@ -50,104 +32,227 @@
+		addClasspathPatternToScan("classpath*:messages*.properties");
+		addDomainToScan(InitializerConstants.DOMAIN_ADDR);
+		addDomainToScan(InitializerConstants.DOMAIN_MSGPROP);
+		addFallbackLanguage("ht", "fr");
```
Wait, is the idea that we are hardcoding these fallbacks into the message source itself? That doesn't seem correct.

mseaton replied:
My idea is that we would provide a means to add fallback languages, but that yes, we would specifically make a decision that - by default - French is the fallback for Haitian Kreyol for message localization.  This feels right to me.

mseaton replied:
OK, talked this over with @mogoodrich and agreed to remove this hard-coded fallback in favor of letting this be configured as-needed.

mogoodrich commented on api/src/main/java/org/openmrs/module/initializer/InitializerMessageSource.java:
```diff
@@ -50,104 +32,227 @@
+ * the best match in the System Locale. If no match is found in the System Locale, then the message
+ * code itself is returned as a final fallback. This source allows for supporting additional
+ * fallback languages, as well as defining additional classpath patterns and domains to search for
+ * message property files. See method javadoc for further details.
```
I would reference the method name here, ie "addFallbackLanguage"

mseaton replied:
Done

mogoodrich reviewed (changes requested):
Generally looks good, one small comment regarding hardcoding the Kreyol fallback (probably something you forgot to clean up)?

mseaton commented:
@mks-d - FYI, I have just created a copy of this class and added it into one of our PIH modules, and wired it up as the main message source.  This will give us some opportunities to test this out in the wild with our actual distribution running across various countries and test servers, and provide better information as to how this works beyond any code review that I fear will not otherwise allow us to reach the level of confidence needed to merge this in.  Let's circle back around this in a few days and see how things stand and then aim to get this merged in if tests are successful?  Please review in the meantime and let me know if anything looks concerning.

mks-d reviewed (commented):
>I have just created a copy of this class and added it into one of our PIH modules, and wired it up as the main message source. This will give us some opportunities to test this out in the wild with our actual distribution running across various countries and test servers

@mseaton I wonder if your in-house testing on the copy of this enhanced class are positive? If yes I'd be keen to merge this in.

mseaton commented:
> > I have just created a copy of this class and added it into one of our PIH modules, and wired it up as the main message source. This will give us some opportunities to test this out in the wild with our actual distribution running across various countries and test servers
> 
> @mseaton I wonder if your in-house testing on the copy of this enhanced class are positive? If yes I'd be keen to merge this in.

@mks-d yes they have been positive.  I've just updated the PR to more accurately reflect our in-house version, which had a few changes made to it since the initial copy.  It was pretty much unchanged, with the exception being how it is setup initially.  I found that just relying on the ApplicationContextAware handling was not reliable enough in terms of ordering.  Instead, I changed it to be a component and then explicitly set it as the active message source in the started method of the activator.  This ensures that it is set after all other earlier modules and core have loaded and that this is done last.

I'd recommend we merge this in and provide an opportunity early enough prior to the next release to allow existing users to test it out, but overall we have been happy with this.
